chainerx_cc/chainerx/routines/logic.cc

```diff
@@ -7,70 +7,32 @@
 #include "chainerx/kernels/logic.h"
 #include "chainerx/routines/creation.h"
 #include "chainerx/routines/manipulation.h"
+#include "chainerx/routines/routines_util.h"
 #include "chainerx/routines/type_util.h"
 #include "chainerx/shape.h"
 
 namespace chainerx {
 
 namespace {
 
-template <typename Impl>
-Array BroadcastComparison(Impl&& impl, const Array& x1, const Array& x2) {
-    auto func = [&impl](const Array& x1, const Array& x2) {
-        Array out = Empty(x1.shape(), Dtype::kBool, x1.device());
-        {
-            NoBackpropModeScope scope{};
-            impl(x1, x2, out);
-        }
-        return out;
-    };
-
-    if (x1.shape() == x2.shape()) {
-        return func(x1, x2);
-    }
-    Shape result_shape = internal::BroadcastShapes(x1.shape(), x2.shape());
-    if (x1.shape() == result_shape) {
-        return func(x1, x2.BroadcastTo(result_shape));
-    }
-    if (x2.shape() == result_shape) {
-        return func(x1.BroadcastTo(result_shape), x2);
-    }
-    return func(x1.BroadcastTo(result_shape), x2.BroadcastTo(result_shape));
-}
-
-void CheckLogicDtypes(const Array& x1, const Array& x2) {
+template <typename Kernel>
+Array LogicBinaryImpl(const Array& x1, const Array& x2) {
     if ((x1.dtype() == Dtype::kBool) != (x2.dtype() == Dtype::kBool)) {
         throw DtypeError{"Comparison of ", GetDtypeName(x1.dtype()), " and ", GetDtypeName(x2.dtype()), " is not supported."};
     }
+    auto func = [](const Array& x1, const Array& x2, Array& out) { x1.device().backend().CallKernel<Kernel>(x1, x2, out); };
+    return internal::BroadcastBinary(func, x1, x2, Dtype::kBool);
```

Comment:
Did you intentionally remove the `NoBackpropModeScope`, since it's not created inside `internal::BroadcastBinary`? I think it's safer to put it back for now.

Reply:
Fixed.